[Bug]: Cannot create QuickSight dataset which uses cast column type operation

### Terraform Core Version

1.4.5

### AWS Provider Version

4.62.0

### Affected Resource(s)

`aws_quicksight_data_set`

### Expected Behavior

Should be able to create QuickSight dataset

### Actual Behavior

Error was returned

### Relevant Error/Panic Output Snippet

```
╷
│ Error: error setting logical_table_map: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}
│
│   with aws_quicksight_data_set.test,
│   on main.tf line 52, in resource "aws_quicksight_data_set" "test":
│   52: resource "aws_quicksight_data_set" "test" {
│
```


### Terraform Configuration Files

```terraform
provider "aws" {
  region = "eu-west-1"
}

resource "random_uuid" "val" {}

resource "aws_s3_bucket" "bucket" {
  bucket = random_uuid.val.result
}

resource "aws_s3_object" "object" {
  bucket = aws_s3_bucket.bucket.id
  key    = "manifest.json"
  content = <<EOT
{
    "fileLocations": [ { "URIs": [] } ],
    "globalUploadSettings": { "format": "JSON" }
}
EOT
}

resource "aws_quicksight_data_source" "s3" {
  data_source_id = "example-id"
  name           = "S3"
  type = "S3"

  parameters {
    s3 {
      manifest_file_location {
        bucket = aws_s3_bucket.bucket.id
        key    = "manifest.json"
      }
    }
  }
}

resource "aws_quicksight_data_set" "test" {

  data_set_id    = "test"
  import_mode    = "SPICE"
  name           = "Test"

  physical_table_map {
    physical_table_map_id = "id"
    s3_source {
      data_source_arn = aws_quicksight_data_source.s3.arn
      input_columns {
        name = "var"
        type = "STRING"
      }
      upload_settings {
        format = "JSON"
      }
    }
  }

  logical_table_map {
    logical_table_map_id = "id"
    alias = "alias"
    source {
      physical_table_id = "id"
    }

    data_transforms {
      cast_column_type_operation {
        column_name     = "var"
        new_column_type = "DATETIME"
        format          = "yyyy/MM/dd/HH"
      }
    }
  }
}
```

### Steps to Reproduce

Plan and apply the supplied Terraform configuration.

### Debug Output

```
aws_quicksight_data_set.test: Creating...
2023-04-13T15:37:49.413+0100 [INFO]  Starting apply for aws_quicksight_data_set.test
2023-04-13T15:37:49.413+0100 [DEBUG] aws_quicksight_data_set.test: applying the planned Create change
2023-04-13T15:37:49.414+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "data_set_usage_configuration" from ComputedKeys
2023-04-13T15:37:49.414+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "tags_all" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.source.0.join_instruction" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.rename_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.tag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.untag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.create_columns_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.filter_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.project_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.project_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.rename_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.tag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.untag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.create_columns_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.filter_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.rename_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.tag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.untag_column_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.create_columns_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.filter_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.data_transforms.0.project_operation" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.source.0.join_instruction" from ComputedKeys
2023-04-13T15:37:49.415+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: [DEBUG] setting computed for "logical_table_map.0.source.0.join_instruction" from ComputedKeys
2023-04-13T15:37:49.416+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: HTTP Request Sent: http.request_content_length=423 http.user_agent="APN/1.0 HashiCorp/1.0 Terraform/1.4.5 (+https://www.terraform.io) terraform-provider-aws/4.62.0 (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go/1.44.237 (go1.19.7; linux; amd64)" tf_mux_provider=*schema.GRPCProviderServer tf_provider_addr=registry.terraform.io/hashicorp/aws aws.operation=CreateDataSet aws.sdk=aws-sdk-go aws.service=QuickSight http.request.header.x_amz_date=20230413T143749Z tf_resource_type=aws_quicksight_data_set tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2@v2.0.0-beta.26/logger.go:90 net.peer.name=quicksight.eu-west-1.amazonaws.com @module=aws http.request.header.content_type=application/json http.request.header.x_amz_security_token=***** http.url=https://quicksight.eu-west-1.amazonaws.com/accounts/ACCOUNT_ID/data-sets http.request.body={"DataSetId":"test","ImportMode":"SPICE","LogicalTableMap":{"id":{"Alias":"alias","DataTransforms":[{"CastColumnTypeOperation":{"ColumnName":"var","Format":"yyyy/MM/dd/HH","NewColumnType":"DATETIME"}}],"Source":{"PhysicalTableId":"id"}}},"Name":"Test","PhysicalTableMap":{"id":{"S3Source":{"DataSourceArn":"arn:aws:quicksight:eu-west-1:ACCOUNT_ID:datasource/example-id","InputColumns":[{"Name":"var","Type":"STRING"}]}}}} http.request.header.authorization="AWS4-HMAC-SHA256 Credential=ASIA************Z6RO/20230413/eu-west-1/quicksight/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-security-token, Signature=*****" tf_req_id=2b5aab99-933a-9803-ff0c-8f80201ebafb aws.region=eu-west-1 http.flavor=1.1 http.method=POST timestamp=2023-04-13T15:37:49.416+0100
2023-04-13T15:37:50.553+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: HTTP Response Received: http.status_code=201 tf_mux_provider=*schema.GRPCProviderServer http.duration=1137 aws.region=eu-west-1 http.response.body={"Arn":"arn:aws:quicksight:eu-west-1:ACCOUNT_ID:dataset/test","DataSetId":"test","IngestionArn":"arn:aws:quicksight:eu-west-1:ACCOUNT_ID:dataset/test/ingestion/662aea0b-7456-4684-8e79-5424087b9bc4","IngestionId":"662aea0b-7456-4684-8e79-5424087b9bc4","RequestId":"0fab773f-34c9-43e8-a454-f150db16a3fd"} tf_provider_addr=registry.terraform.io/hashicorp/aws tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2@v2.0.0-beta.26/logger.go:138 aws.operation=CreateDataSet http.response.header.x_amzn_requestid=0fab773f-34c9-43e8-a454-f150db16a3fd tf_req_id=2b5aab99-933a-9803-ff0c-8f80201ebafb tf_resource_type=aws_quicksight_data_set aws.sdk=aws-sdk-go http.response.header.content_type=application/json http.response.header.date="Thu, 13 Apr 2023 14:37:50 GMT" @module=aws aws.service=QuickSight timestamp=2023-04-13T15:37:50.553+0100
2023-04-13T15:37:50.554+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: HTTP Request Sent: http.request.header.x_amz_date=20230413T143750Z http.url=https://quicksight.eu-west-1.amazonaws.com/accounts/ACCOUNT_ID/data-sets/test tf_mux_provider=*schema.GRPCProviderServer tf_resource_type=aws_quicksight_data_set @caller=github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2@v2.0.0-beta.26/logger.go:90 @module=aws aws.sdk=aws-sdk-go http.flavor=1.1 http.request.header.authorization="AWS4-HMAC-SHA256 Credential=ASIA************Z6RO/20230413/eu-west-1/quicksight/aws4_request, SignedHeaders=host;x-amz-date;x-amz-security-token, Signature=*****" http.request.header.x_amz_security_token=***** aws.operation=DescribeDataSet aws.service=QuickSight http.request.body= net.peer.name=quicksight.eu-west-1.amazonaws.com tf_rpc=ApplyResourceChange aws.region=eu-west-1 http.method=GET http.user_agent="APN/1.0 HashiCorp/1.0 Terraform/1.4.5 (+https://www.terraform.io) terraform-provider-aws/4.62.0 (+https://registry.terraform.io/providers/hashicorp/aws) aws-sdk-go/1.44.237 (go1.19.7; linux; amd64)" tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=2b5aab99-933a-9803-ff0c-8f80201ebafb timestamp=2023-04-13T15:37:50.554+0100
2023-04-13T15:37:51.080+0100 [DEBUG] provider.terraform-provider-aws_v4.62.0_x5: HTTP Response Received: http.response.header.content_type=application/json tf_mux_provider=*schema.GRPCProviderServer http.duration=526 http.response.header.x_amzn_requestid=f11c7114-9151-452b-b31a-15988c1809c6 tf_provider_addr=registry.terraform.io/hashicorp/aws tf_req_id=2b5aab99-933a-9803-ff0c-8f80201ebafb aws.operation=DescribeDataSet aws.service=QuickSight http.response.header.date="Thu, 13 Apr 2023 14:37:51 GMT" http.status_code=200 tf_resource_type=aws_quicksight_data_set @module=aws aws.region=eu-west-1 aws.sdk=aws-sdk-go http.response.body={"DataSet":{"Arn":"arn:aws:quicksight:eu-west-1:ACCOUNT_ID:dataset/test","ColumnGroups":null,"ColumnLevelPermissionRules":null,"ConsumedSpiceCapacityInBytes":0,"CreatedTime":1.681396670111E9,"DataSetId":"test","DataSetUsageConfiguration":{"DisableUseAsDirectQuerySource":false,"DisableUseAsImportedSource":false},"DatasetParameters":null,"FieldFolders":null,"ImportMode":"SPICE","LastUpdatedTime":1.681396670111E9,"LogicalTableMap":{"id":{"Alias":"alias","DataTransforms":[{"CastColumnTypeOperation":{"ColumnName":"var","Format":"yyyy/MM/dd/HH","NewColumnType":"DATETIME"},"CreateColumnsOperation":null,"FilterOperation":null,"OverrideDatasetParameterOperation":null,"ProjectOperation":null,"RenameColumnOperation":null,"TagColumnOperation":null,"UntagColumnOperation":null}],"Source":{"DataSetArn":null,"JoinInstruction":null,"PhysicalTableId":"id"}}},"Name":"Test","OutputColumns":[{"Description":null,"Name":"var","Type":"DATETIME"}],"PhysicalTableMap":{"id":{"CustomSql":null,"RelationalTable":null,"S3Source":{"DataSourceArn":"arn:aws:quicksight:eu-west-1:ACCOUNT_ID:datasource/example-id","InputColumns":[{"Name":"var","Type":"STRING"}],"UploadSettings":{"ContainsHeader":true,"Delimiter":null,"Format":"JSON","StartFromRow":null,"TextQualifier":"DOUBLE_QUOTE"}}}},"RowLevelPermissionDataSet":null,"RowLevelPermissionTagConfiguration":null},"RequestId":"f11c7114-9151-452b-b31a-15988c1809c6"} tf_rpc=ApplyResourceChange @caller=github.com/hashicorp/aws-sdk-go-base/v2/awsv1shim/v2@v2.0.0-beta.26/logger.go:138 timestamp=2023-04-13T15:37:51.080+0100
2023-04-13T15:37:51.080+0100 [ERROR] provider.terraform-provider-aws_v4.62.0_x5: [ERROR] setting state: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}
2023-04-13T15:37:51.081+0100 [ERROR] provider.terraform-provider-aws_v4.62.0_x5: Response contains error diagnostic: tf_proto_version=5.3 tf_req_id=2b5aab99-933a-9803-ff0c-8f80201ebafb tf_rpc=ApplyResourceChange diagnostic_severity=ERROR @module=sdk.proto diagnostic_detail= diagnostic_summary="error setting logical_table_map: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}" tf_provider_addr=registry.terraform.io/hashicorp/aws tf_resource_type=aws_quicksight_data_set @caller=github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/diag/diagnostics.go:55 timestamp=2023-04-13T15:37:51.081+0100
2023-04-13T15:37:51.098+0100 [ERROR] vertex "aws_quicksight_data_set.test" error: error setting logical_table_map: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}
╷
│ Error: error setting logical_table_map: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}
│
│   with aws_quicksight_data_set.test,
│   on main.tf line 52, in resource "aws_quicksight_data_set" "test":
│   52: resource "aws_quicksight_data_set" "test" {
│
╵
2023-04-13T15:37:51.105+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-04-13T15:37:51.113+0100 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/4.62.0/linux_amd64/terraform-provider-aws_v4.62.0_x5 pid=92031
2023-04-13T15:37:51.113+0100 [DEBUG] provider: plugin exited
```

### Panic Output

_No response_

### Important Factoids

The AWS QuickSight resource **is** created successfully however it appears that updating the TF state is what fails.

Subsequent attempts to perform a plan also fail with the following error message:

```
terraform plan -out plan
random_uuid.val: Refreshing state... [id=0c79c024-7ebb-d260-2f8b-8630b1dfc45d]
aws_s3_bucket.bucket: Refreshing state... [id=0c79c024-7ebb-d260-2f8b-8630b1dfc45d]
aws_s3_object.object: Refreshing state... [id=manifest.json]
aws_quicksight_data_source.s3: Refreshing state... [id=ACCOUNT_ID/example-id]
aws_quicksight_data_set.test: Refreshing state... [id=ACCOUNT_ID,test]

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: error setting logical_table_map: Invalid address to set: []string{"logical_table_map", "2846344263", "data_transforms", "0", "cast_column_type_operation", "0", "cast_column_type_operation"}
│
│   with aws_quicksight_data_set.test,
│   on main.tf line 52, in resource "aws_quicksight_data_set" "test":
│   52: resource "aws_quicksight_data_set" "test" {
│
╵
```

The Terraform state file is as follows:

```json
{
  "version": 4,
  "terraform_version": "1.4.5",
  "serial": 38,
  "lineage": "cb7d28e5-44c1-410a-0941-7b2cea8844f8",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "aws_quicksight_data_set",
      "name": "test",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "status": "tainted",
          "schema_version": 0,
          "attributes": {
            "arn": "arn:aws:quicksight:eu-west-1:ACCOUNT_ID:dataset/test",
            "aws_account_id": "ACCOUNT_ID",
            "column_groups": [],
            "column_level_permission_rules": [],
            "data_set_id": "test",
            "data_set_usage_configuration": [
              {
                "disable_use_as_direct_query_source": false,
                "disable_use_as_imported_source": false
              }
            ],
            "field_folders": [],
            "id": "ACCOUNT_ID,test",
            "import_mode": "SPICE",
            "logical_table_map": [
              {
                "alias": "alias",
                "data_transforms": [
                  {
                    "cast_column_type_operation": [
                      {
                        "column_name": "var",
                        "format": "yyyy/MM/dd/HH",
                        "new_column_type": "DATETIME"
                      }
                    ],
                    "create_columns_operation": [],
                    "filter_operation": [],
                    "project_operation": [],
                    "rename_column_operation": [],
                    "tag_column_operation": [],
                    "untag_column_operation": []
                  }
                ],
                "logical_table_map_id": "id",
                "source": [
                  {
                    "data_set_arn": "",
                    "join_instruction": [],
                    "physical_table_id": "id"
                  }
                ]
              }
            ],
            "name": "Test",
            "permissions": [],
            "physical_table_map": [
              {
                "custom_sql": [],
                "physical_table_map_id": "id",
                "relational_table": [],
                "s3_source": [
                  {
                    "data_source_arn": "arn:aws:quicksight:eu-west-1:ACCOUNT_ID:datasource/example-id",
                    "input_columns": [
                      {
                        "name": "var",
                        "type": "STRING"
                      }
                    ],
                    "upload_settings": [
                      {
                        "contains_header": false,
                        "delimiter": "",
                        "format": "JSON",
                        "start_from_row": 0,
                        "text_qualifier": ""
                      }
                    ]
                  }
                ]
              }
            ],
            "row_level_permission_data_set": [],
            "row_level_permission_tag_configuration": [],
            "tags": null,
            "tags_all": null
          },
          "sensitive_attributes": [],
          "private": "bnVsbA==",
          "dependencies": [
            "aws_quicksight_data_source.s3"
          ]
        }
      ]
    },
    {
      "mode": "managed",
      "type": "aws_quicksight_data_source",
      "name": "s3",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "arn": "arn:aws:quicksight:eu-west-1:ACCOUNT_ID:datasource/example-id",
            "aws_account_id": "ACCOUNT_ID",
            "credentials": [],
            "data_source_id": "example-id",
            "id": "ACCOUNT_ID/example-id",
            "name": "S3",
            "parameters": [
              {
                "amazon_elasticsearch": [],
                "athena": [],
                "aurora": [],
                "aurora_postgresql": [],
                "aws_iot_analytics": [],
                "jira": [],
                "maria_db": [],
                "mysql": [],
                "oracle": [],
                "postgresql": [],
                "presto": [],
                "rds": [],
                "redshift": [],
                "s3": [
                  {
                    "manifest_file_location": [
                      {
                        "bucket": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
                        "key": "manifest.json"
                      }
                    ]
                  }
                ],
                "service_now": [],
                "snowflake": [],
                "spark": [],
                "sql_server": [],
                "teradata": [],
                "twitter": []
              }
            ],
            "permission": [],
            "ssl_properties": [],
            "tags": {},
            "tags_all": {},
            "type": "S3",
            "vpc_connection_properties": []
          },
          "sensitive_attributes": [],
          "private": "bnVsbA==",
          "dependencies": [
            "aws_s3_bucket.bucket",
            "random_uuid.val"
          ]
        }
      ]
    },
    {
      "mode": "managed",
      "type": "aws_s3_bucket",
      "name": "bucket",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "acceleration_status": "",
            "acl": null,
            "arn": "arn:aws:s3:::0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
            "bucket": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
            "bucket_domain_name": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d.s3.amazonaws.com",
            "bucket_prefix": "",
            "bucket_regional_domain_name": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d.s3.eu-west-1.amazonaws.com",
            "cors_rule": [],
            "force_destroy": false,
            "grant": [
              {
                "id": "852142ab362e14beddc4ff4ce50f2799863b45e36c5012353def7ec742a85c93",
                "permissions": [
                  "FULL_CONTROL"
                ],
                "type": "CanonicalUser",
                "uri": ""
              }
            ],
            "hosted_zone_id": "Z1BKCTXD74EZPE",
            "id": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
            "lifecycle_rule": [],
            "logging": [],
            "object_lock_configuration": [],
            "object_lock_enabled": false,
            "policy": "",
            "region": "eu-west-1",
            "replication_configuration": [],
            "request_payer": "BucketOwner",
            "server_side_encryption_configuration": [
              {
                "rule": [
                  {
                    "apply_server_side_encryption_by_default": [
                      {
                        "kms_master_key_id": "",
                        "sse_algorithm": "AES256"
                      }
                    ],
                    "bucket_key_enabled": false
                  }
                ]
              }
            ],
            "tags": {},
            "tags_all": {},
            "timeouts": null,
            "versioning": [
              {
                "enabled": false,
                "mfa_delete": false
              }
            ],
            "website": [],
            "website_domain": null,
            "website_endpoint": null
          },
          "sensitive_attributes": [],
          "private": "eyJlMmJmYjczMC1lY2FhLTExZTYtOGY4OC0zNDM2M2JjN2M0YzAiOnsiY3JlYXRlIjoxMjAwMDAwMDAwMDAwLCJkZWxldGUiOjM2MDAwMDAwMDAwMDAsInJlYWQiOjEyMDAwMDAwMDAwMDAsInVwZGF0ZSI6MTIwMDAwMDAwMDAwMH19",
          "dependencies": [
            "random_uuid.val"
          ]
        }
      ]
    },
    {
      "mode": "managed",
      "type": "aws_s3_object",
      "name": "object",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "acl": "private",
            "bucket": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
            "bucket_key_enabled": false,
            "cache_control": "",
            "content": "{\n    \"fileLocations\": [ { \"URIs\": [] } ],\n    \"globalUploadSettings\": { \"format\": \"JSON\" }\n}\n",
            "content_base64": null,
            "content_disposition": "",
            "content_encoding": "",
            "content_language": "",
            "content_type": "binary/octet-stream",
            "etag": "4b08369cef41be53c636bbf20cc11210",
            "force_destroy": false,
            "id": "manifest.json",
            "key": "manifest.json",
            "kms_key_id": null,
            "metadata": {},
            "object_lock_legal_hold_status": "",
            "object_lock_mode": "",
            "object_lock_retain_until_date": "",
            "server_side_encryption": "AES256",
            "source": null,
            "source_hash": null,
            "storage_class": "STANDARD",
            "tags": {},
            "tags_all": {},
            "version_id": "",
            "website_redirect": ""
          },
          "sensitive_attributes": [],
          "private": "bnVsbA==",
          "dependencies": [
            "aws_s3_bucket.bucket",
            "random_uuid.val"
          ]
        }
      ]
    },
    {
      "mode": "managed",
      "type": "random_uuid",
      "name": "val",
      "provider": "provider[\"registry.terraform.io/hashicorp/random\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d",
            "keepers": null,
            "result": "0c79c024-7ebb-d260-2f8b-8630b1dfc45d"
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    }
  ],
  "check_results": null
}
```

### References

https://github.com/hashicorp/terraform-provider-aws/pull/30349

### Would you like to implement a fix?

None

--- a/.changelog/30701.txt
+++ b/.changelog/30701.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_quicksight_data_set: Fix to properly flatten `cast_column_type_operation.format` field
+```

--- a/internal/service/quicksight/data_set.go
+++ b/internal/service/quicksight/data_set.go
@@ -1976,7 +1976,7 @@ func flattenCastColumnTypeOperation(apiObject *quicksight.CastColumnTypeOperatio
 		tfMap["column_name"] = aws.StringValue(apiObject.ColumnName)
 	}
 	if apiObject.Format != nil {
-		tfMap["cast_column_type_operation"] = aws.StringValue(apiObject.ColumnName)
+		tfMap["format"] = aws.StringValue(apiObject.Format)
 	}
 	if apiObject.NewColumnType != nil {
 		tfMap["new_column_type"] = aws.StringValue(apiObject.NewColumnType)

